### PR TITLE
Release v0.9.0

### PR DESCRIPTION
**Pull request by z4y4ts**
### Added

- Direct transcripts streaming from 3PlayMedia service.
  Now Video Xblock will always use up-to-date transcripts from 3PM.

### Changed

- UI: Transcripts settings section now split into two panels.
  To make more apparent how different options are related and which
  transcripts are going to be diplayed.
  - "Manual & default transcripts"
  - "3PlayMedia transcripts"

### CODE REVIEW

File: video_xblock/mixins.py [hunk 1]
```diff
@@ -165,86 +195,98 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+            transcript = self.fetch_single_3pm_translation(transcript_data)
+            if transcript is None:
+                raise StopIteration
+            transcript_ordered_dict = transcript._asdict()
+            transcript_ordered_dict['content'] = ''  # we don't want to parse it to JSON
+            yield transcript_ordered_dict
 
-    @XBlock.json_handler
-    def get_transcripts_3playmedia_api_handler(self, data, _suffix=''):
+    def get_3pm_transcripts_list(self, file_id, apikey):
```
Comment (z4y4ts) [inline video_xblock/mixins.py:221]: Here we fetch up to date list of available translations for a given File ID and serve it back to the Video player.

File: video_xblock/mixins.py [hunk 2]
```diff
@@ -165,86 +195,98 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+
+        if response.ok and isinstance(response.json(), list):
+            transcripts_list = response.json()
+            feedback['status'] = Status.success
+            feedback['message'] = success_message
+        else:
+            feedback['status'] = Status.error
+        return feedback, transcripts_list
+
+    def fetch_single_3pm_translation(self, transcript_data, format_id=TPMApiTranscriptFormatID.WEBVTT):
```
Comment (z4y4ts) [inline video_xblock/mixins.py:252]: Here we fetch up to date transcript from 3PlayMedia and serve it to the video player.

File: video_xblock/backends/vimeo.py [hunk 3]
```diff
@@ -51,7 +51,7 @@ def get(self, url, headers=None, can_retry=False):
             Response in python native data format.
         """
         headers_ = {
-            'Authorization': 'Bearer {}'.format(self.access_token),
+            'Authorization': 'Bearer {}'.format(self.access_token.encode(encoding='utf-8')),
```
Comment (bryanlandia) [inline video_xblock/backends/vimeo.py:54]: Is this fix something that is important enough to put in the changelog?

File: video_xblock/backends/brightcove.py [hunk 4]
```diff
@@ -497,13 +504,15 @@ def authenticate_api(self, **kwargs):
             error_status_message (str): Error messages for the sake of verbosity.
         """
         token, account_id = kwargs.get('token'), kwargs.get('account_id')
-        client_secret, client_id, error_message = BrightcoveApiClient.create_credentials(token, account_id)
+        try:
+            client_secret, client_id, error_message = BrightcoveApiClient.create_credentials(token, account_id)
+        except BrightcoveApiClientError as bc_exception:
+            return {}, bc_exception.message
```
Comment (bryanlandia) [inline video_xblock/backends/brightcove.py:510]: Is this fix important enough to put in the changelog?

File: video_xblock/backends/vimeo.py [hunk 5]
```diff
@@ -218,7 +225,7 @@ def get_default_transcripts(self, **kwargs):
         try:
             json_data = self.api_client.get(url)
         except VimeoApiClientError:
-            message = _('No timed transcript may be fetched from a video platform.')
+            message = _('No timed transcript may be fetched from a video platform.<br>')
```
Comment (bryanlandia) [inline video_xblock/backends/vimeo.py:228]: `<br/>` :)

File: video_xblock/mixins.py [hunk 6]
```diff
@@ -165,86 +195,98 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
         """
-        Xblock handler to authenticate to a video platform's API. Called by JavaScript of `studio_view`.
+        Make API request to fetch list of available transcripts for given file ID.
 
-        Arguments:
-            data (dict): Data from frontend, necessary for authentication (tokens, account id, etc).
-            suffix (str): Slug used for routing.
-        Returns:
-            response (dict): Status messages key-value pairs.
+        :return: (list of dicts OR dict) all available transcripts attached to file with ID OR error dict
```
Comment (bryanlandia) [inline video_xblock/mixins.py:225]: It looks like it returns a list of dicts AND a feedback dict

File: video_xblock/mixins.py [hunk 7]
```diff
@@ -285,6 +327,67 @@ def srt_to_vtt(self, request, _suffix=''):
...
+            webob.Response: WebVTT transcripts wrapped in Response object.
+        """
+        lang_id, transcript_id = request.query_string.split('=')
+        transcript = self.fetch_single_3pm_translation(transcript_data={'id': transcript_id, 'language_id': lang_id})
+        if transcript is None:
+            return Response()
+        return Response(transcript.content)
+
+    @XBlock.handler
+    def validate_three_play_media_config(self, request, _suffix=''):
```
Comment (bryanlandia) [inline video_xblock/mixins.py:349]: I like that you are validating on setting streaming enabled

File: video_xblock/static/html/studio-edit-3playmedia-apikey.html [hunk 8]
```diff
@@ -1,18 +1,14 @@
 {% load i18n %}
 
 <div class="custom-field-list">
-  <div class="custom-field-section-label">{% trans "3playmedia Key" %}</div>
+  <div class="custom-field-section-label">
+    {% trans "3playmedia Key" %}
```
Comment (bryanlandia) [inline video_xblock/static/html/studio-edit-3playmedia-apikey.html:5]: Tiny thing, but to be consistent, "3PlayMedia Key"

File: video_xblock/static/html/studio-edit-transcripts-accordion.html [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+          <span class="tip setting-help"> {{ field.help|safe }} </span>
+        {% endif %}
+      </li>
+    {% endfor %}
+    {#% include 'studio-edit-3playmedia-apikey.html' %#}
+
+  </div>
+</div>
+
+<script type="text/javascript">
```
Comment (bryanlandia) [inline video_xblock/static/html/studio-edit-transcripts-accordion.html:45]: Great you do most of this in plain JS!

File: video_xblock/tests/unit/test_backends.py [hunk 10]
```diff
@@ -494,7 +512,7 @@ def test_get_default_transcripts_get_failed(self):
         """
 
         # Arrange
-        failure_message = _('No timed transcript may be fetched from a video platform.')
+        failure_message = _('No timed transcript may be fetched from a video platform.<br>')
```
Comment (bryanlandia) [inline video_xblock/tests/unit/test_backends.py:515]: `<br/>`

File: video_xblock/mixins.py [hunk 11]
```diff
@@ -58,6 +56,31 @@ class TranscriptsMixin(XBlock):
     TranscriptsMixin class to encapsulate transcripts-related logic.
     """
 
+    THREE_PLAY_MEDIA_API_DOMAIN = 'https://static.3playmedia.com/'
+
+    threeplaymedia_streaming = Boolean(
+        default=False,
+        display_name=_('Direct 3PlayMedia'),
+        scope=Scope.content,
+        help=_("Direct <a href='http://www.3playmedia.com/'>3PlayMedia</a> transcripts usage enabled.")
```
Comment (bryanlandia) [inline video_xblock/mixins.py:65]: Maybe clarify this, like  `"Enable direct streaming of transcripts from <a href='http://www.3playmedia.com/'>3PlayMedia</a> servers."`

File: video_xblock/backends/vimeo.py [hunk 12]
```diff
@@ -126,7 +126,14 @@ def advanced_fields(self):
...
+        return super(VimeoPlayer, self).advanced_fields
+
+    @property
+    def three_pm_fields(self):
+        """
+        Tuple of VideoXBlock fields to display on `3PlayMedia transcripts` panel.
+        """
+        fields_list = super(VimeoPlayer, self).three_pm_fields
         # Add `token` field before `threeplaymedia_file_id`
         fields_list.insert(fields_list.index('threeplaymedia_file_id'), 'token')
```
Comment (bryanlandia) [inline video_xblock/backends/vimeo.py:138]: This doesn't appear to be working correctly.  I only see the field name and help text, but no HTML form fields under token.  See https://github.com/raccoongang/xblock-video/issues/260

**Review by bryanlandia — APPROVED** [review]
Overall, the new 3PM functionality, accordion organization, and refactoring look excellent! 
Looks like there is 1 regression with the token field, and I've filed an issue in RG github for it.  

Approving and merging, but please look at the token issue, and if you can make some minor text changes from my comments in an 0.9.1 that would be great.

**Pull request merged**